### ISSUE TITLE
Nov 2024 - Upgrade third-party composer dependencies

### ISSUE THREAD

**Issue by emteknetnz** [silverstripe/.github#306]
The CMS 6 beta will be released around mid Jan 2025. Upgrading the composer dependencies should be done fairly close to this cut-off to allow as much time for third party new major versions to be released, though still allow some time to ensure we have adequate time to upgrade everything.

## Acceptance criteria
- [x] Dependencies that need to be updated for CMS 6 are identified (excluding symfony and phpunit)
- [x] Dependencies are updated
- [x] Fixed dependencies page is updated

## Notes
- Symfony 7 has already been done in [this card](https://github.com/silverstripe/.github/issues/296)
- PHPunit 11 has already been done in [this card](https://github.com/silverstripe/.github/issues/305)

This bash script can be used to identify dependencies that need to be updated on the base folder of an install of kitchen-sink:

The following script was used to update deps - https://github.com/emteknetnz/cms6-update-deps

### Kitchen sink CI
- https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/12225931067

^ Unit test failure is [existing](https://github.com/silverstripe/silverstripe-framework/actions/runs/12228337036/job/34106455689)

### PRs
- https://github.com/silverstripe/silverstripe-assets/pull/666
- https://github.com/silverstripe/silverstripe-behat-extension/pull/293
- https://github.com/silverstripe/silverstripe-crontask/pull/98
- https://github.com/silverstripe/silverstripe-dynamodb/pull/71
- https://github.com/silverstripe/silverstripe-event-dispatcher/pull/32
- https://github.com/silverstripe/silverstripe-framework/pull/11498
- https://github.com/silverstripe/silverstripe-frameworktest/pull/216
- https://github.com/silverstripe/silverstripe-graphql/pull/617
- https://github.com/silverstripe/silverstripe-gridfieldqueuedexport/pull/130
- https://github.com/silverstripe/silverstripe-mfa/pull/580
- https://github.com/silverstripe/MinkFacebookWebDriver/pull/31
- https://github.com/silverstripe/silverstripe-realme/pull/173
- https://github.com/silverstripe/recipe-testing/pull/30
- https://github.com/silverstripe/silverstripe-textextraction/pull/105
- https://github.com/silverstripe/silverstripe-totp-authenticator/pull/179
- https://github.com/silverstripe/documentation-lint/pull/12
- https://github.com/silverstripe/markdown-php-codesniffer/pull/16
- https://github.com/silverstripe/MinkFacebookWebDriver/pull/32
- https://github.com/silverstripe/documentation-lint/pull/13

After merging assign back to Steve 

## Update major deps

Did not update any phpunit deps as they were all for repos that are used for both CMS 5 and CMS 6 e.g. silverstripe/supported-modules

### Kitchen sink CI
- https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/12268029139

^ Unit test failure is [existing](https://github.com/silverstripe/silverstripe-framework/actions/runs/12228337036/job/34106455689)

### PRs
- https://github.com/silverstripe/silverstripe-framework/pull/11505
- https://github.com/silverstripe/silverstripe-totp-authenticator/pull/180
- https://github.com/silverstripe/developer-docs/pull/653

**Comment by GuySartorelli** [comment]
Got a printout of all the dependencies per repo, then realised this isn't actually worth my time until we've got the list of supported modules.

Used this script, modified from above:

```php
$deps = [];
$vendors = ["silverstripe","symbiote","tractorcow","colymba","dnadesign","bringyourownideas"];
foreach ($vendors as $vendor) {
    $files = shell_exec("find vendor/$vendor/. | grep composer.json");
    $files = explode("\n", $files);
    foreach ($files as $file) {
        if (!$file) continue;
        if (preg_match("#/tests/#", $file)) continue;
        preg_match("#vendor/(?<repo>.*)/composer\.json$#", $file, $matches);
        $repo = $matches["repo"];
        $json = json_decode(file_get_contents($file), true);
        foreach (["require", "require-dev"] as $attr) {
            if (!isset($json[$attr])) continue;
            foreach ($json[$attr] as $dep => $constraint) {
                if ($dep === "php") continue;
                if (str_starts_with($dep, "silverstripe/")) continue;
                if (str_starts_with($dep, "phpunit/")) continue;
                if (str_starts_with($dep, "bringyourownideas/")) continue;
                if (str_starts_with($dep, "colymba/")) continue;
                if (str_starts_with($dep, "dnadesign/")) continue;
                if (str_starts_with($dep, "symbiote/")) continue;
                if (str_starts_with($dep, "symfony/")) continue;
                if (str_starts_with($dep, "ext-")) continue;
                $repoDeps[$repo][] = "$dep:$constraint";
            }
        }
    }
}
ksort($repoDeps);
foreach ($repoDeps as $repo => $deps) {
  ksort($deps);
  echo "\n---------------\n";
  echo "REPO: $repo\n---------------\n";
  foreach ($deps as $dep) {
    echo "$dep\n";
  }
}
```

**Comment by emteknetnz** [comment]
Made minor update to my original script and ran, and manually got the latest version from each repo

Only 2 things are not on the latest major:
- paragonie/constant_time_encoding - ^2.6 - latest is 3.0.0
- psr/http-message - ^1 - latest is 2.0 (not a typo)

```
asyncphp/doorman:^4
- 4.0.0
aws/aws-sdk-php:^3.281
- 3.334.1
behat/behat:^3.11.0
- 3.16.0
behat/mink:^1.10.0
- 1.12.0
composer-plugin-api:^2
- keep
composer-runtime-api:^2.0
- keep
composer/composer:^2
- 2.8.3
ccomposer/installers:*
- 2.3.0
composer/installers:^2
- 2.3.0
composer/installers:^2.2
- 2.3.0
composer/semver:^3.4
- 3.4.3
defuse/php-encryption:^2.3
- 2.4.0
dragonmantank/cron-expression:^3
- 3.4.0
embed/embed:^4.4.7
- 4.4.14
friends-of-behat/mink-extension:^2
- 2.7.5
fzaninotto/faker:^1.9.2
- 1.9.2
guzzlehttp/guzzle:^7.5
- 7.9.2
guzzlehttp/guzzle:^7.5.0
- 7.9.2
guzzlehttp/psr7:^2.4.0
- 2.7.0
guzzlehttp/psr7:^2.4.1
- 2.7.0
intervention/image:^3.7
- 3.9.1
jeremeamia/superclosure:^2.0
- 2.4.0
league/commonmark:^2.4
- 2.5.3
league/csv:^9
- 9.18.0
league/csv:^9.8.0
- 9.18.0
league/flysystem-local:^3.22
- 3.29.0
league/flysystem:^3.22
- 3.29.1
m1/env:^2.2.0
- 2.2.0
marcj/topsort:^2.0.0
- 2.0.0
masterminds/html5:^2.7.6
- 2.9.0
mikey179/vfsstream:^1.6
- 1.6.12
mikey179/vfsstream:^1.6.11
- 1.6.12
mikey179/vfsstream:^v1.6.11
- 1.6.12
monolog/monolog:^3.2.0
- 3.8.0
nikic/php-parser:^5.1.0
- 5.3.1
onelogin/php-saml:^4
- 4.2.0
paragonie/constant_time_encoding:^2.6
- ! 3.0.0
php-parallel-lint/php-parallel-lint:^1
- 1.4.0
php-webdriver/webdriver:^1.13.1
- 1.15.2
php:^8.0
- ^8.3
php:^8.1
- ^8.3
php:^8.3
- ^8.3
phpstan/extension-installer:^1.3
- 1.4.3
phpunit/phpunit:^11.3
- 11.4.4
phpunit/phpunit:^4.0 || ^5.0 || ^6.0 || ^7.0 || ^8.0 || ^9.0
- 11.4.4
phpunit/phpunit:^9.6
- 11.4.4
psr/container:^1.1 || ^2.0
- 2.0.2
psr/event-dispatcher:^1
- 1.0.0
psr/http-message:^1
- ! 2.0
psr/simple-cache:^3.0.0
- 3.0.0
sebastian/diff:^6.0
- 6.0.2
sensiolabs/ansi-to-html:^1.2
- 1.2.1
silverstripe-themes/simple:3.x-dev
- 3.3.2
slevomat/coding-standard:^8.14
- 8.15.0
sminnee/callbacklist:^0.1.1
- 0.1.1
spomky-labs/otphp:^11.1
- 11.3.0
squizlabs/php_codesniffer:^3
- 3.7.2
squizlabs/php_codesniffer:^3.7
- 3.7.2
ua-parser/uap-php:^3.9.14
- 3.9.14
webonyx/graphql-php:^15.0.1
- 15.19.0
```

**Comment by emteknetnz** [comment]
@GuySartorelli 3 options:

## Option A
- No changes exception to try and get paragonie/constant_time_encoding and psr/http-message to next major
- Least amount of work
- Widest compatibility
- Highest chance of projects not getting important patches

## Option B
- Update all deps to latest minor e.g. intervention/image:^3.7 => ^3.9.1
- Restricted compatibility
- Low chance of projects not getting important patches

## Option C
- Update all deps to latest patch e.g. intervention/image:^3.7 => ^3.9.1
- Most restricted compatibility
- Lowest chance of projects not getting important patches
- Maybe a little confusing looking at composer.json in future cos would assume we needed patch to fix specific bug, which is not the case
- Does not match how we have done deps in the past

I'm personally not a fan of the "widest compatibility" idea with option A, where it does make the upgrade to CMS 6 easier if there is something other third-party dep that only works with some old version of the dep, but at the cost of not getting potentially important bugfixes / security fixes, and it also widens our maintenance surface.

Option C is probably overkill as no one installs with --prefer lowest in the real world

My preference is we go with option B - do you have any objection to that

**Comment by GuySartorelli** [comment]
Definitely not C.
I prefer A for compatibility but in reality almost everything uses caret constraints so the chance of compatibility problems with option B is very low.

Tl;dr either A or B is fine 👍

**Comment by emteknetnz** [comment]
Cool with go with B

**Comment by GuySartorelli** [comment]
@emteknetnz I'm getting this when trying to build CMS 5 now:

```text
Problem 1
    - silverstripe/recipe-testing[3.0.0, ..., 3.x-dev] require silverstripe/behat-extension ^5 -> satisfiable by silverstripe/behat-extension[5.0.0, ..., 5.x-dev].
    - silverstripe/behat-extension 5.x-dev is an alias of silverstripe/behat-extension dev-master and thus requires it to be installed too.
    - silverstripe/behat-extension[5.0.0, ..., 5.x-dev] require silverstripe/mink-facebook-web-driver ^2 -> satisfiable by silverstripe/mink-facebook-web-driver[2.x-dev].
    - silverstripe/behat-extension dev-master requires php ^7.2 -> your php version (8.1; overridden via config.platform, actual: 8.3.14) does not satisfy that requirement.
    - silverstripe/mink-facebook-web-driver 2.x-dev requires php ^8.3 -> your php version (8.1; overridden via config.platform, actual: 8.3.14) does not satisfy that requirement.
    - Root composer.json requires silverstripe/recipe-testing ^3 -> satisfiable by silverstripe/recipe-testing[3.0.0, ..., 3.x-dev].
```

I think the wrong branch may have been used for recipe testing, behat extension, or the mink one.

**Comment by emteknetnz** [comment]
Likely mink-facebook webdriver - have made pr https://github.com/silverstripe/MinkFacebookWebDriver/pull/32

**Comment by GuySartorelli** [comment]
PRs merged - reassigned to Steve for next steps

**Comment by GuySartorelli** [comment]
PRs merged
